### PR TITLE
Fix for PHP 8.0 when symlink function is disabled

### PR DESCRIPTION
**Pull request by mvorisek**
see https://3v4l.org/FJZNm

`@` operator in PHP 8.0 does NOT suppress fatal error thrown when a disabled function is called, thus we need to check explicitly if disabled first. `@` operator continues to hide php warnings thus the operator before the file function should stay.

This fix is fully BC.

```
Fatal error: Uncaught Error: Call to undefined function Composer\Util\symlink() in phar:///usr/local/bin/composer/src/Composer/Util/Filesystem.php:604
Stack trace:
#0 phar:///usr/local/bin/composer/src/Composer/Installer/BinaryInstaller.php(160): Composer\Util\Filesystem->relativeSymlink('/builds/mvorise...', '/builds/mvorise...')
#1 phar:///usr/local/bin/composer/src/Composer/Installer/BinaryInstaller.php(88): Composer\Installer\BinaryInstaller->installSymlinkBinaries('/builds/mvorise...', '/builds/mvorise...')
#2 phar:///usr/local/bin/composer/src/Composer/Installer/LibraryInstaller.php(143): Composer\Installer\BinaryInstaller->installBinaries(Object(Composer\Package\CompletePackage), '/builds/mvorise...')
#3 [internal function]: Composer\Installer\LibraryInstaller->Composer\Installer\{closure}(NULL)
#4 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/FulfilledPromise.php(20): call_user_func(Object(Closure), NULL)
#5 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred.php(74): React\Promise\FulfilledPromise->then(Object(Closure), NULL)
#6 [internal function]: React\Promise\Deferred->React\Promise\{closure}(Object(React\Promise\FulfilledPromise))
#7 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred.php(166): call_user_func(Object(Closure), Object(React\Promise\FulfilledPromise))
#8 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred.php(95): React\Promise\Deferred->processQueue(Array, Object(React\Promise\FulfilledPromise))
#9 [internal function]: React\Promise\Deferred->resolve(NULL)
#10 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/FulfilledPromise.php(20): call_user_func(Array, NULL)
#11 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred.php(76): React\Promise\FulfilledPromise->then(Array, Array, Array)
#12 [internal function]: React\Promise\Deferred->React\Promise\{closure}(Object(React\Promise\FulfilledPromise))
#13 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred.php(166): call_user_func(Object(Closure), Object(React\Promise\FulfilledPromise))
#14 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred.php(95): React\Promise\Deferred->processQueue(Array, Object(React\Promise\FulfilledPromise))
#15 [internal function]: React\Promise\Deferred->resolve(NULL)
#16 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/FulfilledPromise.php(20): call_user_func(Array, NULL)
#17 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred.php(76): React\Promise\FulfilledPromise->then(Array, Array, Array)
#18 [internal function]: React\Promise\Deferred->React\Promise\{closure}(Object(React\Promise\FulfilledPromise))
#19 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred.php(166): call_user_func(Object(Closure), Object(React\Promise\FulfilledPromise))
#20 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred.php(95): React\Promise\Deferred->processQueue(Array, Object(React\Promise\FulfilledPromise))
#21 [internal function]: React\Promise\Deferred->resolve(Object(Symfony\Component\Process\Process))
#22 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/FulfilledPromise.php(20): call_user_func(Array, Object(Symfony\Component\Process\Process))
#23 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred.php(76): React\Promise\FulfilledPromise->then(Array, Array, Array)
#24 [internal function]: React\Promise\Deferred->React\Promise\{closure}(Object(React\Promise\FulfilledPromise))
#25 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred.php(166): call_user_func(Object(Closure), Object(React\Promise\FulfilledPromise))
#26 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Deferred.php(95): React\Promise\Deferred->processQueue(Array, Object(React\Promise\FulfilledPromise))
#27 phar:///usr/local/bin/composer/vendor/react/promise/src/React/Promise/Promise.php(42): React\Promise\Deferred->resolve(Object(Symfony\Component\Process\Process))
#28 [internal function]: React\Promise\Promise->React\Promise\{closure}(Object(Symfony\Component\Process\Process))
#29 phar:///usr/local/bin/composer/src/Composer/Util/ProcessExecutor.php(284): call_user_func(Object(Closure), Object(Symfony\Component\Process\Process))
#30 phar:///usr/local/bin/composer/src/Composer/Util/Loop.php(95): Composer\Util\ProcessExecutor->countActiveJobs()
#31 phar:///usr/local/bin/composer/src/Composer/Installer/InstallationManager.php(449): Composer\Util\Loop->wait(Array, NULL)
#32 phar:///usr/local/bin/composer/src/Composer/Installer/InstallationManager.php(435): Composer\Installer\InstallationManager->waitOnPromises(Array)
#33 phar:///usr/local/bin/composer/src/Composer/Installer/InstallationManager.php(362): Composer\Installer\InstallationManager->executeBatch(Object(Composer\Repository\InstalledFilesystemRepository), Array, Array, true, true, Array)
#34 phar:///usr/local/bin/composer/src/Composer/Installer/InstallationManager.php(261): Composer\Installer\InstallationManager->downloadAndExecuteBatch(Object(Composer\Repository\InstalledFilesystemRepository), Array, Array, true, true, Array)
#35 phar:///usr/local/bin/composer/src/Composer/Installer.php(721): Composer\Installer\InstallationManager->execute(Object(Composer\Repository\InstalledFilesystemRepository), Array, true, true)
#36 phar:///usr/local/bin/composer/src/Composer/Installer.php(555): Composer\Installer->doInstall(Object(Composer\Repository\InstalledFilesystemRepository), true)
#37 phar:///usr/local/bin/composer/src/Composer/Installer.php(251): Composer\Installer->doUpdate(Object(Composer\Repository\InstalledFilesystemRepository), true)
#38 phar:///usr/local/bin/composer/src/Composer/Command/UpdateCommand.php(230): Composer\Installer->run()
#39 phar:///usr/local/bin/composer/vendor/symfony/console/Command/Command.php(245): Composer\Command\UpdateCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#40 phar:///usr/local/bin/composer/vendor/symfony/console/Application.php(835): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#41 phar:///usr/local/bin/composer/vendor/symfony/console/Application.php(185): Symfony\Component\Console\Application->doRunCommand(Object(Composer\Command\UpdateCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#42 phar:///usr/local/bin/composer/src/Composer/Console/Application.php(310): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#43 phar:///usr/local/bin/composer/vendor/symfony/console/Application.php(117): Composer\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#44 phar:///usr/local/bin/composer/src/Composer/Console/Application.php(122): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#45 phar:///usr/local/bin/composer/bin/composer(63): Composer\Console\Application->run()
#46 /usr/local/bin/composer(24): require('phar:///usr/loc...')
#47 {main}
```

### CODE REVIEW

File: src/Composer/Util/Filesystem.php [hunk 1]
```diff
@@ -601,7 +601,12 @@ public function relativeSymlink($target, $link)
 
         $relativePath = $this->findShortestPath($link, $target);
         chdir(\dirname($link));
-        $result = @symlink($relativePath, $link);
+        
+        if (function_exists('symlink')) {
+            $result = @symlink($relativePath, $link);
+        } else {
+            $result = false;
```
Comment (mvorisek) [inline src/Composer/Util/Filesystem.php:608]: for max. compatibility, can be moved to the method beginning, advise/edit

File: src/Composer/Util/Filesystem.php [hunk 2]
```diff
@@ -597,6 +597,10 @@ private function unlinkImplementation($path)
      */
     public function relativeSymlink($target, $link)
     {
+        if (!function_exists('symlink')) {
```
Comment (mvorisek) [inline src/Composer/Util/Filesystem.php:600]: 2nd commit is less BC, but I think we do not need to chdir and back

**Comment by GrahamCampbell** [comment]
Why is there still an `@` on `symlink`. What other kind of failures are we expecting?

**Comment by GrahamCampbell** [comment]
Isn't the issue here that we need a try-catch around symlink. `@` suppresses "errors" but not "Errors".

**Comment by mvorisek** [comment]
@GrahamCampbell I address a different issue.

I belive in this PR we should keep `@` to supress php warnings like when symlink was not created due no space or RO filesystem. Try catch will not help as php file functions does not throw.

**Comment by GrahamCampbell** [comment]
The try catch will catch the case when the function is not defined, any any other such errors that throw an Error rather than triggering an error.

**Comment by mvorisek** [comment]
@GrahamCampbell no, `symlink` (and other file functions) does NEVER throw except when not defined - https://3v4l.org/K74vq

I belive, the proposed fix fixes the issue correctly and as I have written, try catch will still not supress possible warnings from these functions.

If you still belive you have a better solution, please submit a PR against my source branch of this PR.

**Comment by GrahamCampbell** [comment]
They do throw Errors:

image

https://3v4l.org/UhNup

**Comment by GrahamCampbell** [comment]
Your example also shows that PHP 8 throws errors for disabled functions, unlike PHP 7:

image

**Comment by mvorisek** [comment]
@GrahamCampbell what you want to achive? stop spamming

In my opinion, `function_exists` handles the only possible sceneraio when "try catch might be used" and I do NOT feel catching general `Error` is better than testing `function_exists`

**Comment by GrahamCampbell** [comment]
My point was the OP does not correctly explain why this change is being made:

> @ operator in php8 does NOT suppress fatal errors anymore

The reason for the issue is that calls to disabled functions have been reclassified from raising a warning to throwing an Error. This OP is misleading to reviewers.

**Comment by GrahamCampbell** [comment]
No, that's still incorrect. It does not **emit** a **fatal error**. it **throws** an **Error**. Uncaught throwables are then rendered as fatal errors, and the `@` has **never** suppressed stuff that's thrown.

**Comment by mvorisek** [comment]
@GrahamCampbell please post the exact description you want me to have here.

**Comment by Seldaek** [comment]
The fix seems appropriate to me so I'll merge. Thanks.

**Pull request merged**